Device/ARM/ARMCR52/startup_ARMCR52.c

```diff
@@ -0,0 +1,159 @@
+void EL2_Reserved_Handler (void) __attribute__ ((weak, alias("EL2_Default_Handler")));
+void EL2_IRQ_Handler      (void) __attribute__ ((weak, alias("EL2_Default_Handler")));
+void EL2_FIQ_Handler      (void) __attribute__ ((weak, alias("EL2_Default_Handler")));
+
+void EL1_Reset_Handler    (void) __attribute__ ((weak, alias("EL1_Default_Handler")));
+void EL1_Undef_Handler    (void) __attribute__ ((weak, alias("EL1_Default_Handler")));
+void EL1_SVC_Handler      (void) __attribute__ ((weak, alias("EL1_Default_Handler")));
+void EL1_Prefetch_Handler (void) __attribute__ ((weak, alias("EL1_Default_Handler")));
+void EL1_Abort_Handler    (void) __attribute__ ((weak, alias("EL1_Default_Handler")));
+void EL1_Reserved_Handler (void) __attribute__ ((weak, alias("EL1_Default_Handler")));
+void EL1_IRQ_Handler      (void) __attribute__ ((weak, alias("EL1_Default_Handler")));
+void EL1_FIQ_Handler      (void) __attribute__ ((weak, alias("EL1_Default_Handler")));
+
+/*----------------------------------------------------------------------------
+  Exception / Interrupt Vector table
+ *----------------------------------------------------------------------------*/
+void __VECTOR_TABLE_ATTRIBUTE __EL2_Vectors(void) {
+  __ASM volatile(
+  "LDR    PC,=EL2_Reset_Handler  \n"
+  "B.W    EL2_Undef_Handler      \n"
+  "B.W    EL2_SVC_Handler        \n"
+  "B.W    EL2_Prefetch_Handler   \n"
+  "B.W    EL2_Abort_Handler      \n"
+  "B.W    EL2_Reserved_Handler   \n"
+  "B.W    EL2_IRQ_Handler        \n"
+  "B.W    EL2_FIQ_Handler        \n"
+  );
+}
+
+void __VECTOR_TABLE_ATTRIBUTE __EL1_Vectors(void) {
+  __ASM volatile(
+  "B.W    EL1_Reset_Handler      \n"
+  "B.W    EL1_Undef_Handler      \n"
+  "B.W    EL1_SVC_Handler        \n"
+  "B.W    EL1_Prefetch_Handler   \n"
+  "B.W    EL1_Abort_Handler      \n"
+  "B.W    EL1_Reserved_Handler   \n"
+  "B.W    EL1_IRQ_Handler        \n"
+  "B.W    EL1_FIQ_Handler        \n"
+  );
+}
+
+
+/*----------------------------------------------------------------------------
+  Switch from EL2 to EL1 mode
+ *----------------------------------------------------------------------------*/
+void __attribute__((naked,__noreturn__)) SwitchEL2toEL1(uint32_t func) {
+  register CPSR_Type spsr_hyp;
+
+  __set_ELR_Hyp(func);
+
+  spsr_hyp.w = __get_SPSR_Hyp();
+#if defined(__thumb2__) && (__thumb2__==1)
+  /* Thumb instruction set and supervisor mode */
+  spsr_hyp.b.M = CPSR_M_SVC;
+  spsr_hyp.b.T = 1;
+#else
+  /* ARM instruction set and supervisor mode */
+  spsr_hyp.b.M = CPSR_M_SVC;
+#endif
+  __set_SPSR_Hyp(spsr_hyp.w);
+
+  __ERET();
+
+  __builtin_unreachable(); /* No return here */
+}
+
+/*----------------------------------------------------------------------------
+  Reset Handler called on controller reset
+ *----------------------------------------------------------------------------*/
+void __attribute__((naked,__noreturn__)) EL2_Reset_Handler(void) {
```

Comment:
Naked functions implemented in C? At least in ArmClang this won't work, I suspect.

Reply:
We cant rely on that stack won't be used here since SP isn't initialized yet (no hardware SP setup as in M profile). I've seen already that ``register`` keyword doesn't guarantee this for you.

Reply:
Yes, indeed. I leave the choice up to you whether you want to investigate in a generalized startup code (like I did it for M-class recently) or stay with compiler specific files (like it is for A-class) for the time being. Compiler specific code should go into a sub directory.